src/IceRpc.Slice.Generators/Internal/Parser.cs

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) ZeroC, Inc.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace IceRpc.Slice.Generators.Internal;
+
+internal sealed class Parser
+{
+    internal const string OperationAttribute = "IceRpc.Slice.SliceOperationAttribute";
+    internal const string ServiceAttribute = "IceRpc.Slice.SliceServiceAttribute";
+
+    private readonly CancellationToken _cancellationToken;
+    private readonly Compilation _compilation;
+    private readonly INamedTypeSymbol? _operationAttribute;
+    private readonly Action<Diagnostic> _reportDiagnostic;
+    private readonly INamedTypeSymbol? _serviceAttribute;
+
+    internal Parser(
+        Compilation compilation,
+        Action<Diagnostic> reportDiagnostic,
+        CancellationToken cancellationToken)
+    {
+        _compilation = compilation;
+        _reportDiagnostic = reportDiagnostic;
+        _cancellationToken = cancellationToken;
+
+        _operationAttribute = _compilation.GetTypeByMetadataName(OperationAttribute);
+        _serviceAttribute = _compilation.GetTypeByMetadataName(ServiceAttribute);
+    }
+
+    internal IReadOnlyList<ServiceClass> GetServiceDefinitions(IEnumerable<ClassDeclarationSyntax> classes)
+    {
+        if (_operationAttribute is null || _serviceAttribute is null)
+        {
+            // nothing to do if these types aren't available
+            return Array.Empty<ServiceClass>();
+        }
+
+        var serviceDefinitions = new List<ServiceClass>();
+        // we enumerate by syntax tree, to minimize the need to instantiate semantic models (since they're expensive)
+        foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax> group in classes.GroupBy(x => x.SyntaxTree))
+        {
+            foreach (ClassDeclarationSyntax classDeclaration in group)
+            {
+                // stop if we're asked to
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                SemanticModel semanticModel = _compilation.GetSemanticModel(classDeclaration.SyntaxTree);
+                INamedTypeSymbol? classSymbol = semanticModel.GetDeclaredSymbol(classDeclaration, _cancellationToken);
+                if (classSymbol is null)
+                {
+                    continue;
+                }
+
+                IReadOnlyList<ServiceMethod> baseServiceMethods = Array.Empty<ServiceMethod>();
+                bool hasBaseServiceDefinition = false;
+                if (classSymbol.BaseType is INamedTypeSymbol baseType)
```

Comment:
The logic does not seem correct to me. Take:
```csharp
[SliceService] internal partial class MyBase : IBase { ... }

internal class MyDerived : MyBase, IDerived { ... } // no SliceService attribute, so no code generation

[SliceService] internal partial class MyMoreDerived : MyDerived, IMoreDerived { ...}
```

I believe you'll generate code that skips the dispatching to operations on Derived.

The `baseServiceMethods` that you filter out must correspond to the methods on the nearest base class where you found the SliceService attribute.

Would be good to add a test!

Reply:
Added a test